### ISSUE TITLE
SVG should specify what CSS attribute selectors and class selectors match on

### ISSUE THREAD

**Issue by birtles** [w3c/svgwg#328]
Background: [Mozilla bug 1365472](https://bugzilla.mozilla.org/show_bug.cgi?id=1365472#c12)

Given that attributes can be animated in SVG, SVG needs to specify if attribute selectors match on the animated value of the attribute or not. Similarly for class selectors.

Proposal:
* Attribute selectors match on the base value of the attribute, i.e. the same value as `getAttribute` returns since [SMIL Animation](https://www.w3.org/TR/2001/REC-smil-animation-20010904/#BasicAnim) says, "When an animation is running, it should not actually change the attribute values in the DOM." (Note that Chrome appears to have [a bug](https://bugs.chromium.org/p/chromium/issues/detail?id=735820) here.)
* Class selectors match on the set of classes applied to the element which can be affected by SMIL animation (as otherwise being able to animate the `class` attribute would be meaningless, and being able to animate this is quite useful).

That means that if we have:

```svg
<circle>
  <set attributeName="class" to="a"/>
</circle>
```

Then`circle.a` would match but `circle[class="a"]` would not.

cc @bzbarsky

**Comment by tabatkins** [comment]
That sounds... quite bad? We've never broken the symmetry between class and attr selectors before.

**Comment by AmeliaBR** [comment]
Class and attribute selectors aren't perfectly symmetrical, given that class selectors are processed (tokenized). Even if you use the token form of the attribute selector, I'm sure there could be weird edge cases where you get inconsistent results.  

In general, attribute selectors always match the literal attribute values, even if those are not valid values for the corresponding DOM properties, so you can get disconnects that way.  For example, `lang` selectors or other pseudoclasses could be out of sync with querying the actual attributes they are based on.

But that said: 

I would be quite happy to move towards an interpretation of SVG animation as more a declarative way of changing attributes, so the actual attribute value would change while the animation was in effect (just like it does if you're animating with JS).  In other words, if you `<set>` an attribute on an element, treat it as if it has that attribute for selector-matching.

But I think that goes against the argument for dropping `animVal` from SVG 2 (and I think some browsers already): there are DOM costs for keeping everything in sync.

**Comment by tabatkins** [comment]
> Even if you use the token form of the attribute selector, I'm sure there could be weird edge cases where you get inconsistent results.

No, using `[class ~= "a"]` should work *precisely* the same as `.a`.

**Comment by bzbarsky** [comment]
> We've never broken the symmetry between class and attr selectors before

They have different case-sensitivity behavior in quirks mode, for what it's worth.

But also, the behavior @birtles proposes is what Safari and Firefox ship already.  Edge doesn't support SMIL, and Chrome has broken `[class="foo"]` matching; see https://bugs.chromium.org/p/chromium/issues/detail?id=735820 (used to match Safari and Firefox before that bug was introduced).  So all browsers that support SMIL already match on the animated value for class selectors.

Conceptually, in spec terms as things stand, the "set of classes an element belongs to" is not necessarily a function of any of its attributes at all; it needs to be defined by relevant specifications that define that element.

**Comment by tabatkins** [comment]
Yeah, that's true. Makes me a little sad, but my sadness is worth nothing in comparison to everyone agreeing.

**Comment by boggydigital** [comment]
Not blocking updated 2.0 CR publication - assigning 2.1 WD milestone
